keadm/cmd/keadm/app/cmd/helm/cloudcore.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package helm
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kubeedge/kubeedge/common/constants"
+	types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	kecharts "github.com/kubeedge/kubeedge/manifests"
+)
+
+const (
+	CloudCoreHelmComponent = "cloudcore"
+	ChartsSubDir           = "charts"
+
+	setsKeyCloudcoreImageTag         = "cloudCore.image.tag"
+	setsKeyIptablesManagerImageTag   = "iptablesManager.image.tag"
+	setsKeyControllerManagerImageTag = "controllerManager.image.tag"
+
+	MessageFormatInstallationSuccess = `%s started
+=========CHART DETAILS=======
+Name: %s
+LAST DEPLOYED: %s
+NAMESPACE: %s
+STATUS: %s
+REVISION: %d
+`
+
+	MessageFormatUpgradationPrintConfig = `This is cloudcore configuration of the previous version.
+If you want to revert configuration items, please manually modify the configmap 'cloudcore' and restart the cloudcore: \n%s`
+)
+
+// CloudcoreHelmTool a cloudcore helm chart operation wrapping tool
+type CloudcoreHelmTool struct {
+	util.Common
+}
+
+// NewCloudcoreHelmTool creates a new instance of CloudcoreHelmTool
+func NewCloudcoreHelmTool(kubeConfig, kubeedgeVersion string) *CloudcoreHelmTool {
+	common := util.Common{
+		ToolVersion:     semver.MustParse(util.GetHelmVersion(kubeedgeVersion, util.RetryTimes)),
+		KubeConfig:      kubeConfig,
+		OSTypeInstaller: util.GetOSInterface(),
+	}
+	common.OSTypeInstaller.SetKubeEdgeVersion(common.ToolVersion)
+	return &CloudcoreHelmTool{
+		Common: common,
+	}
+}
+
+// Install uses helm client to install cloudcore release
+func (c *CloudcoreHelmTool) Install(opts *types.InitOptions) error {
+	// The flag --force would not care about whether the cloud components exist or not also.
+	// If gives a external helm root, no need to check and verify, because it is always not a cloudcore.
```

Comment:
I'm not sure what the purpose of utilizing ExternalHelmRoot was before, but it was not in the development process this time. You can keep it in the installation procedure but exclude it from the upgrade process.

Reply:
Ok, this function is used to abstract the helm related operations of the cloudcore, and the code of the cloud init has not been modified.